rocAL/source/readers/file_source_reader.cpp

```diff
@@ -41,10 +42,32 @@ FileSourceReader::FileSourceReader() {
 }
 
 unsigned FileSourceReader::count_items() {
-    if (_loop)
-        return _file_names.size();
-
-    int ret = ((int)_file_names.size() - _read_counter);
+    int ret = 0;
+    if (_shard_size == -1) {                                     // When shard_size is set to -1, The shard_size variable is not used
+        if (_loop) return largest_shard_size_without_padding();  // Return the size of the largest shard amongst all the shard's size
```

Comment:
Can't we use dataset_size here?

Reply:
consider using better formula for computing shard_size like below.
floor((id + 1) * dataset_size / num_shards) -
    floor(id * dataset_size / num_shards)
Also for Fill Vs Drop
You can use ceil(shard_size/batch_size) * batch_size and floor(shard_size/batch_size)*batch_size

Reply:
The formula to calculate actual shard_size without padding used in file_source_reader.cpp is
size_t FileSourceReader::actual_shard_size_without_padding() {
    return std::floor((_shard_id + 1) * get_dataset_size() / _shard_count) - std::floor(_shard_id * get_dataset_size() / _shard_count);
}
which is picked from DALI https://docs.nvidia.com/deeplearning/dali/user-guide/docs/advanced_topics_sharding.html#shard-calculation